Before you turn this problem in, make sure everything runs as expected. First, **restart the kernel** (in the menubar, select Kernel$\rightarrow$Restart) and then **run all cells** (in the menubar, select Cell$\rightarrow$Run All).

Make sure you fill in any place that says `YOUR CODE HERE` or "YOUR ANSWER HERE", as well as your name and collaborators below:

In [1]:
NAME = "Senna Renting"
STUDENT_NUMBER = "s1067489"
COLLABORATOR_NAME = ""
COLLABORATOR_STUDENT_NUMBER = ""

---

## **Important:** When handing in your homework:
+ Hand in the notebook (and nothing else) **named as follows**: StudentName1_snumber_StudentName2_snumber.ipynb
+ Provide clear and complete answers to the questions below under a separate header (not hidden somewhere in your source code), and make sure to explain your answers / motivate your choices. Add Markdown cells where necessary.
+ Source code, output graphs, derivations, etc., should be included in the notebook.
+ Hand-in: upload to Brightspace.
+ Include name, student number, assignment (especially in filenames)!
+ When working in pairs only one of you should upload the assignment, and report the name of your partner in your filename.
+ Use the Brightspace discussion board or email the student assistants for questions on how to complete the exercises.
+ If you find mistakes/have suggestions/would like to complain about the assigment material itself, please email me [Roel] at `Roel.Bouman@ru.nl`
+ Do not remove any cells in the notebook, else this might break the auto-grading system. **An invalid notebook will mean a severe reduction in your grade!**
+ Many online collaboration platforms remove metadata from notebooks, which breaks the auto-grading system. Again, **An invalid notebook will mean a severe reduction in your grade!**. Should you wish to use these platforms, copy the answers from the online notebook to one running on your own machine with Anaconda, and then execute all cells.
+ Only type your answers in those places where they are asked.
+ Remove any "raise NotImplementedError()" statements in the cells you answered.

# Assignment 5: Association mining

## Objective of this assignment
The overall objective is to understand how frequent itemsets can be extracted by
the Apriori algorithm and be able to calculate and interpret association rules in terms of support and confidence.


## Advised Reading and Exercise Material
**The following reading material is recommended:**

- Pang-Ning Tan, Michael Steinbach, and Vipin Kumar, *Introduction to Data Mining*, section 6.


## Additional Tools
For this exercise you will need to load the provided *apriorimining.py* script. 


##  5.1 Association mining for course data 
We will use the Apriori algorithm to automatically mine for associations. The Apriori algorithm is an adapted version of the script found here: https://github.com/nalinaksh/Association-Rule-Mining-Python

Check out the script and doc and check if you understand how the association rules are computed. 




#### 5.1.1

(0.25 points) Look at the data file `data/courses.txt` by **using Python**. The data is represented in Table 1. Inspect the file `data/courses.txt` and make sure you understand how the data in Table 1 is stored in the text file. Explain how the table is a representation of the text file. What information that we would need for complete reconstruction of the table is missing from the text file?

##### Table 1
|#  |   History |Math| Biology| Spanish | Economics| Physics | Chemistry | English  |  
| :-------------: |:-------------:| :-----------:| :----------:| :------------:|:-------------:| :------------:|  :-------------: | :-------------: |
|student 1 | 0| 1 | 0 | 0 | 1| 1 |1 |1   
|student 2 | 1| 1 | 1 | 0 | 0| 1 |1 |1   
|student 3 | 0| 1 | 0 | 1 | 0| 1 |0 |1   
|student 4 | 0| 0 | 1 | 0 | 0| 1 |1 |0   
|student 5 | 0| 1 | 0 | 0 | 0| 1 |1 |0        
|student 6 | 0| 1 | 1 | 0 | 0| 1 |1 |1   


In [2]:
##Answer to question 5.1.1
filename = "./data/courses.txt"
with open(filename, "r") as f:
    print("Text file courses.txt:")
    print(f.read())

Text file courses.txt:
2,5,6,7,8
1,2,3,6,7,8
2,4,6,8
3,6,7
2,6,7
2,3,6,7,8



**Answer:** The column names are missing in the text file. Additionally the attributes that contain zeroes are left out in the text file. So in order to understand the data from the text file, we would either need a specific line with column names, or we need a README that explains how the text file is structured in terms of it's attributes.

#### 5.1.2
(0.75 points) We will analyze the data in Table 1  using the function `apriorimining.generate_association_rules()` from the toolbox. This function takes the a string dictating the file path as input. Analyze the data with $ minsupport  \geq 80 \% $ and $ minconfidence \geq 100 \%$. (input as fractions into the function)
What are the generated association rules? What kind of conclusions can you make based on these association rules about the subjects that students took? 
You can optionally provide a dictionary `names`, translating the integers in the file to the names in the table, to the `generate_association_rules` function. 


In [3]:
##Answer to question 5.1.2
# YOUR CODE HERE
from toolbox import apriorimining
apriorimining.generate_association_rules(filename, 0.8, 1.0)

---------------TOP 10 FREQUENT 1-ITEMSET-------------------------
set= { 6 },  sup= 1.0
set= { 2 },  sup= 0.83
set= { 7 },  sup= 0.83
-----------------------------------------------------------------
-------TOP 10 (or less) FREQUENT 2-ITEMSET------------------------
set= { 2, 6 },  sup= 0.83
set= { 6, 7 },  sup= 0.83
------------------------------------------------------------------
---------------------ASSOCIATION RULES------------------
--------------------------------------------------------
Rule #1: {  } ==> { 6 }, sup= 1.00, conf= 1.00

Rule #2: { 2 } ==> { 6 }, sup= 0.83, conf= 1.00

Rule #3: { 7 } ==> { 6 }, sup= 0.83, conf= 1.00
--------------------------------------------------------


**Answers:** All students in this dataset took the physics course, 5 out of 6 took the math course and again 5 out of 6 took the chemistry course. Furthermore when looking at the association rules that were found we see that people who took the math course were very likely (83%) to also take the physics course. In addition to that people who took the chemistry course were also very likely to also take the physics course (83%)

##  5.2 Association mining for MovieLens data 
  
  
In this part of the exercise we consider a Market Basket data set containing 943 users purchases of 1682 movies. A total of 100,000 movies
have been purchased.The data set is called MovieLens100K and is provided by http://www.grouplens.org/node/73, see also the readme `MovieLensData.txt` in the data folder. The data currently considered is not the original data but modified for the apriori algorithm.

#### 5.2.1
(0.25 points) The MovieLens data is stored in the files `data/MovieLensData.txt` and `data/u.item`. Inspect the files **using Python** to see how the data is stored. How do these files relate? 
  

In [4]:
##Answer to question 5.2.1
# YOUR CODE HERE
line_amount = 10
with open("./data/MovieLensData.txt", "r") as f:
    movieData1 = f.readlines()
    movieData1h = "\n".join(movieData1[0:line_amount])
with open("./data/u.item", "r") as f:
    movieData2 = f.readlines()
    movieData2h = "".join(movieData2[0:line_amount])
print("Data from MovieLensData.txt:")
print(movieData1h)
print("Data from u.item:")
print(movieData2h)
print("Length of MovieLensData.txt file vs u.item file respectively: ")
print(f"{len(movieData1)} vs {len(movieData2)}")
print("Try to find last index of u.item file data object in MovieLensData.txt")
last_index = int(movieData2[-1].split("|")[0])
print(f"Last index u.item: {last_index}")
in_movielens = str(last_index) in ("".join(movieData1))
print(f"Is the last index({last_index}) in movieLensData.txt?: {in_movielens}")
more_in_lens = str(last_index+1) in ("".join(movieData1))
print(f"But is there more indices than the u.item file?: {more_in_lens}")

Data from MovieLensData.txt:
1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130,131,132,133,134,135,136,137,138,139,140,141,142,143,144,145,146,147,148,149,150,151,152,153,154,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,172,173,174,175,176,177,178,179,180,181,182,183,184,185,186,187,188,189,190,191,192,193,194,195,196,197,198,199,200,201,202,203,204,205,206,207,208,209,210,211,212,213,214,215,216,217,218,219,220,221,222,223,224,225,226,227,228,229,230,231,232,233,234,235,236,237,238,239,240,241,242,243,244,245,246,247,248,249,250,251,252,253,254,255,256,257,258,259,260,261,262,263,264,265,266,267,268,269,270

**Anwer:** To be honest I can't immediately see, how the two datafiles are related. I first thought each line in the MovieLensData.txt file represented a data object just as the u.item file has, but after checking the amount of lines for both files I quickly found that this couldn't be the case. What did seem possible to me was that each line was an attribute, and that it contained the indices of the u.item if such an attribute was related to any of the data objects in u.item, I checked for the last index of u.item in the MovieLensData.txt file and found this to be the case.

#### 5.2.2 
(0.75 point) Find association rules using the function below with $ minsupport  \geq 30 \% $ and $ minconfidence \geq 80 \%$. You can use the `max_rules` argument to set a maximum number of potential rules to print. Use the `movie_names` dict as the optional `names` argument in `generate_association_rules` to interpret the results. What are the associations with strongest confidence? Do these associations make sense? Explain.

In [5]:
from toolbox.load_movie_names import load_movie_names
movie_names = load_movie_names()
##Answer to question 5.2.2
filename = "./data/MovieLensData.txt"
apriorimining.generate_association_rules(filename, 0.3, 0.8, names=movie_names)

---------------TOP 10 FREQUENT 1-ITEMSET-------------------------
set= { Star Wars (1977) },  sup= 0.62
set= { Fargo (1996) },  sup= 0.54
set= { Return of the Jedi (1983) },  sup= 0.54
set= { Contact (1997) },  sup= 0.54
set= { English Patient, The (1996) },  sup= 0.51
set= { Scream (1996) },  sup= 0.51
set= { Liar Liar (1997) },  sup= 0.51
set= { Toy Story (1995) },  sup= 0.48
set= { Air Force One (1997) },  sup= 0.46
set= { Independence Day (ID4) (1996) },  sup= 0.45
-----------------------------------------------------------------
-------TOP 10 (or less) FREQUENT 2-ITEMSET------------------------
set= { Return of the Jedi (1983), Star Wars (1977) },  sup= 0.51
set= { Fargo (1996), Star Wars (1977) },  sup= 0.42
set= { Star Wars (1977), Toy Story (1995) },  sup= 0.4
set= { Raiders of the Lost Ark (1981), Star Wars (1977) },  sup= 0.4
set= { Independence Day (ID4) (1996), Star Wars (1977) },  sup= 0.38
set= { Godfather, The (1972), Star Wars (1977) },  sup= 0.38
set= { Fargo (1996), R

In general the starwars movies that are present in the purchases, were often bought together, so a lot of association rules is contain them (probably also because they have a high support in this dataset). There were some other interesting rules as well, for example "Toy Story", "Mission: Impossible" and "Pulp Fiction" all were strongly associated with buying "Star Wars". The association rule with the highest confidence (1.0) is: $$\textbf{The Empire Strikes Back + Return of the Jedi + Raiders of the Lost Ark} \rightarrow \textbf{Star Wars}$$. Most of the association rules make sense, but to buy star wars after having bought pulp fiction doesn't really seem like a normal thing to do, so this is probably due to insufficient data, or it is just a weird pattern that people tend to fall into.

#### 5.2.3 
(1 point) Which movies have been watched by the most users? There are only few rules with more than three items. Why?

The top three movies watched by most users are:
1. Star Wars
2. Fargo
3. Return of the Jedi

This is probably due to there being no sequence of movies containing more than 4 items, Also by chance the number of rules decreases for larger itemsets, since the chance of them all being picked by more than just a few users are very unlikely. 

#### 5.2.4
(0.5 points) Often we are interested in rules with high confidence. Is it possible for itemsets to have very low support but still have corresponding rule with a very high confidence?

Yes, because the confidence of an association rule can be retrieved by dividing the itemset containing all items in the rule over just the items that were bought (association items excluded). Support however is measured as the presence of an itemset in the complete dataset (in percentage as well). So because the confidence of an association rule is relative to the transaction itemset, we can easily get a confidence of 1 while having a low support (the explanation can b

## 5.3 Calculating support, confidence and interest

Calculate these measures and write down how you computed things, not just the answers. You can use Latex syntax for writing your answers.


#### 5.3.1
 Suppose we have market basket data consisting of 100 transactions and 20 items. The support for item $ \text{a} = 45 \%$, the support for item $ \text{b} = 80 \%$ and the support for itemset $ \text{ {a,b }} = 30 \%$. Let the support and confidence thresholds be 20$ \%$ and 60$ \%$, respectively.
  
1. (0.5 points) Compute the confidence of the association rule $ \text{ {a } } \rightarrow   \text{{b }} $. Is the rule interesting according to the confidence measure?

2. (0.5 points) Compute the interest measure (or lift, see slide 46 of chapter 6) for the association rule $ \text{ {a } } \rightarrow   \text{{b }} $. Describe the nature of the relationship between item $ \text{a}$ and item $ \text{b}$  in terms of the interest measure.

3. (1 points) What conclusion can you draw from the results of parts (1) and (2)?

YOUR ANSWER HERE

#### 5.3.2

1. (1 points) Let $c_1$, $c_2$, and $c_3$ be the confidence values of the rules $ \text{ {a } } \rightarrow   \text{{b }} $, $ \text{ {a } } \rightarrow   \text{{b,c }} $, and $ \text{ {a,c } } \rightarrow   \text{{b }} $ respectively. If we assume that $c_1$, $c_2$, and $c_3$ have different values, what are the possible inequality relationships (e.g. $c_1 \leq c_2 \leq c_3$) among $c_1$, $c_2$, and $c_3$? Which rule has the lowest confidence?
2. (0.5 points) Suppose the confidence of the rules  $ \text{ {a } } \rightarrow   \text{{b }} $ and $ \text{ {b } } \rightarrow   \text{{c }} $ are larger than the confidence threshold. Is it possible that $ \text{ {a } } \rightarrow   \text{{c }} $ has a confidence below that threshold? If no, explain why. If yes, give an example. 

YOUR ANSWER HERE


#### 5.3.3

(3 points) Consider the relationships between customers who buy high-definition televisions and exercise machines as shown in Table 2 and 3.

1. Compute the odd ratios for both tables.
2. Compute the $\phi$-coefficient for both tables.
3. Compute the interest (or lift, in the book) factor for both tables.

For Table 3 you should compute measures given above separately for College
Students and for Adults. **For each of the measures, describe how the direction
of association changes when data is pooled together** (Table 2) instead of being
separated into two groups (Table 3)

##### Table 2: Two way contingency table between the sale of high-definition television and exercise machine
| |   Buy Exercise machine |     |     |
| :------------- | -------------:| :-----------:| :----------:| 
| **Buy HDTV     ** | yes | no | total |
| yes  | 105| 87 | 192 | 
| no | 40| 62 | 102 |   
| total | 145 | 149 | 294 | 
 

##### Table 3: Example of three-way contingency table
| | |   Buy Exercise machine |     |     |
|--- | :------------- | -------------:| :-----------:| :----------:| 
|**Customer group** | **Buy HDTV     ** | yes | no | total |
|College students | yes  | 2| 9 | 11 | 
| | no | 5| 20 | 25 |
| Working adults | yes  | 103| 78 | 181 | 
| | no | 35| 42 | 77 |  



YOUR ANSWER HERE